File: crates/rivet-repository/src/project_graph.rs
//! # rivet-repository::project_graph
//!
//! Structural Project Graph representation: nodes for files, symbols, modules,
//! build/test targets, and edges carrying explicit provenance.

use rivet_types::*;
use serde::{Deserialize, Serialize};
use std::collections::{HashMap, HashSet};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum NodeKind {
    Repository,
    Directory,
    SourceFile,
    GeneratedFile,
    Symbol,
    Type,
    Module,
    Package,
    BuildTarget,
    TestTarget,
    Service,
    ExternalDependency,
    Config,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum EdgeKind {
    Defines,
    References,
    Calls,
    Imports,
    Inherits,
    Implements,
    Builds,
    LinksTo,
    DependsOn,
    TestedBy,
    Covers,
    Generates,
    ConfiguredBy,
    ServedBy,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct EdgeProvenance {
    pub provider: String,
    pub repo_snapshot: String,
    pub confidence: String,
    pub evidence_refs: Vec<EvidenceId>,
}

impl Default for EdgeProvenance {
    fn default() -> Self {
        Self {
            provider: "deterministic_census".into(),
            repo_snapshot: "r0".into(),
            confidence: "authoritative_within_provider_scope".into(),
            evidence_refs: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectNode {
    pub id: String,
    pub kind: NodeKind,
    pub label: String,
    pub metadata: serde_json::Value,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectEdge {
    pub from: String,
    pub to: String,
    pub kind: EdgeKind,
    pub provenance: EdgeProvenance,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ProjectGraph {
    pub nodes: HashMap<String, ProjectNode>,
    pub edges: Vec<ProjectEdge>,
}

impl ProjectGraph {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn add_node(&mut self, id: impl Into<String>, kind: NodeKind, label: impl Into<String>) {
        let id_str = id.into();
        let label_str = label.into();
        self.nodes.insert(
            id_str.clone(),
            ProjectNode {
                id: id_str,
                kind,
                label: label_str,
                metadata: serde_json::Value::Null,
            },
        );
    }

    pub fn add_node_with_metadata(
        &mut self,
        id: impl Into<String>,
        kind: NodeKind,
        label: impl Into<String>,
        metadata: serde_json::Value,
    ) {
        let id_str = id.into();
        let label_str = label.into();
        self.nodes.insert(
            id_str.clone(),
            ProjectNode {
                id: id_str,
                kind,
                label: label_str,
                metadata,
            },
        );
    }

    pub fn add_edge(
        &mut self,
        from: impl Into<String>,
        to: impl Into<String>,
        kind: EdgeKind,
        provenance: EdgeProvenance,
    ) {
        self.edges.push(ProjectEdge {
            from: from.into(),
            to: to.into(),
            kind,
            provenance,
        });
    }

    pub fn outgoing_edges(&self, node_id: &str) -> Vec<&ProjectEdge> {
        self.edges.iter().filter(|e| e.from == node_id).collect()
    }

    pub fn incoming_edges(&self, node_id: &str) -> Vec<&ProjectEdge> {
        self.edges.iter().filter(|e| e.to == node_id).collect()
    }

    pub fn dependencies_of(&self, node_id: &str) -> Vec<String> {
        self.edges
            .iter()
            .filter(|e| {
                e.from == node_id
                    && matches!(
                        e.kind,
                        EdgeKind::DependsOn
                            | EdgeKind::Imports
                            | EdgeKind::References
                            | EdgeKind::Calls
                    )
            })
            .map(|e| e.to.clone())
            .collect()
    }

    pub fn dependents_of(&self, node_id: &str) -> Vec<String> {
        self.edges
            .iter()
            .filter(|e| {
                e.to == node_id
                    && matches!(
                        e.kind,
                        EdgeKind::DependsOn
                            | EdgeKind::Imports
                            | EdgeKind::References
                            | EdgeKind::Calls
                    )
            })
            .map(|e| e.from.clone())
            .collect()
    }

    pub fn tests_covering(&self, node_id: &str) -> Vec<String> {
        self.edges
            .iter()
            .filter(|e| {
                (e.to == node_id && matches!(e.kind, EdgeKind::Covers | EdgeKind::TestedBy))
                    || (e.from == node_id && e.kind == EdgeKind::TestedBy)
            })
            .map(|e| {
                if e.to == node_id {
                    e.from.clone()
                } else {
                    e.to.clone()
                }
            })
            .collect()
    }

    pub fn symbols_in_file(&self, file_id: &str) -> Vec<String> {
        self.edges
            .iter()
            .filter(|e| e.from == file_id && e.kind == EdgeKind::Defines)
            .map(|e| e.to.clone())
            .collect()
    }

    /// Reachability closure for impact analysis
    pub fn transitive_dependents(&self, root_id: &str) -> HashSet<String> {
        let mut visited = HashSet::new();
        let mut queue = vec![root_id.to_string()];

        while let Some(current) = queue.pop() {
            for dep in self.dependents_of(&current) {
                if visited.insert(dep.clone()) {
                    queue.push(dep);
                }
            }
        }

        visited
    }
}

File: packages/core/src/rivet/repository/project-graph.ts
import { Revision, type EvidenceId } from "../types"

export type NodeKind =
  | "repository"
  | "workspace"
  | "package"
  | "module"
  | "source_file"
  | "generated_file"
  | "symbol"
  | "function"
  | "method"
  | "type"
  | "class"
  | "interface"
  | "struct"
  | "trait"
  | "test"
  | "test_target"
  | "build_target"
  | "service"
  | "external_dependency"
  | "config"
  | "entrypoint"

export type EdgeKind =
  | "contains"
  | "defines"
  | "imports"
  | "exports"
  | "references"
  | "implements"
  | "extends"
  | "inherits"
  | "calls"
  | "builds"
  | "configures"
  | "configured_by"
  | "links_to"
  | "depends_on"
  | "tests"
  | "tested_by"
  | "covers"
  | "generates"
  | "served_by"

export type GraphConfidence =
  | "DETERMINISTIC"
  | "RESOLVED"
  | "SYNTACTIC"
  | "INFERRED"

export interface EdgeProvenance {
  readonly provider: string
  readonly repoSnapshot: string
  readonly confidence: GraphConfidence
  readonly evidenceRefs: readonly EvidenceId[]
}

export interface ProjectNode {
  readonly id: string
  readonly kind: NodeKind
  readonly label: string
  readonly fileId?: string
  readonly metadata?: Record<string, unknown> | null
}

export interface ProjectEdge {
  readonly from: string
  readonly to: string
  readonly kind: EdgeKind
  readonly provenance: EdgeProvenance
}

export interface FileCacheEntry {
  readonly contentHash: string
  readonly parserVersion: string
  readonly indexerVersion: string
  readonly nodeIds: readonly string[]
}

export class ProjectGraph {
  repositoryRevision: Revision
  graphRevision: Revision

  readonly nodes: Map<string, ProjectNode> = new Map()
  readonly edges: ProjectEdge[] = []
  private readonly edgesFromMap: Map<string, ProjectEdge[]> = new Map()
  private readonly edgesToMap: Map<string, ProjectEdge[]> = new Map()
  private readonly fileCache: Map<string, FileCacheEntry> = new Map()

  constructor(repositoryRevision: Revision = Revision.ZERO, graphRevision: Revision = Revision.ZERO) {
    this.repositoryRevision = repositoryRevision
    this.graphRevision = graphRevision
  }

  addNode(
    id: string,
    kind: NodeKind,
    label: string,
    metadata?: Record<string, unknown> | null,
    fileId?: string
  ): void {
    this.nodes.set(id, { id, kind, label, fileId, metadata: metadata ?? null })
  }

  addEdge(
    from: string,
    to: string,
    kind: EdgeKind,
    provenance?: Partial<EdgeProvenance>
  ): void {
    const edge: ProjectEdge = {
      from,
      to,
      kind,
      provenance: {
        provider: provenance?.provider ?? "deterministic_census",
        repoSnapshot: provenance?.repoSnapshot ?? this.repositoryRevision.toString(),
        confidence: provenance?.confidence ?? "DETERMINISTIC",
        evidenceRefs: provenance?.evidenceRefs ?? [],
      },
    }

    this.edges.push(edge)

    const fromList = this.edgesFromMap.get(from) ?? []
    fromList.push(edge)
    this.edgesFromMap.set(from, fromList)

    const toList = this.edgesToMap.get(to) ?? []
    toList.push(edge)
    this.edgesToMap.set(to, toList)
  }

  getNode(id: string): ProjectNode | undefined {
    return this.nodes.get(id)
  }

  outgoingEdges(nodeId: string): ProjectEdge[] {
    return this.edgesFromMap.get(nodeId) ?? []
  }

  incomingEdges(nodeId: string): ProjectEdge[] {
    return this.edgesToMap.get(nodeId) ?? []
  }

  dependenciesOf(nodeId: string): string[] {
    const edges = this.outgoingEdges(nodeId)
    return edges
      .filter(
        (e) =>
          e.kind === "depends_on" ||
          e.kind === "imports" ||
          e.kind === "references" ||
          e.kind === "calls" ||
          e.kind === "implements" ||
          e.kind === "extends" ||
          e.kind === "inherits"
      )
      .map((e) => e.to)
  }

  dependentsOf(nodeId: string): string[] {
    const edges = this.incomingEdges(nodeId)
    return edges
      .filter(
        (e) =>
          e.kind === "depends_on" ||
          e.kind === "imports" ||
          e.kind === "references" ||
          e.kind === "calls" ||
          e.kind === "implements" ||
          e.kind === "extends" ||
          e.kind === "inherits"
      )
      .map((e) => e.from)
  }

  testsCovering(nodeId: string): string[] {
    const list: string[] = []
    for (const e of this.incomingEdges(nodeId)) {
      if (e.kind === "covers" || e.kind === "tested_by" || e.kind === "tests") {
        list.push(e.from)
      }
    }
    for (const e of this.outgoingEdges(nodeId)) {
      if (e.kind === "tested_by" || e.kind === "tests") {
        list.push(e.to)
      }
    }
    return Array.from(new Set(list))
  }

  symbolsInFile(fileId: string): string[] {
    return this.outgoingEdges(fileId)
      .filter((e) => e.kind === "defines" || e.kind === "contains" || e.kind === "exports")
      .map((e) => e.to)
  }

  /**
   * Reachability closure for impact analysis
   */
  transitiveDependents(rootId: string): Set<string> {
    const visited = new Set<string>()
    const queue = [rootId]

    while (queue.length > 0) {
      const current = queue.pop()!
      for (const dep of this.dependentsOf(current)) {
        if (!visited.has(dep)) {
          visited.add(dep)
          queue.push(dep)
        }
      }
    }

    return visited
  }

  /**
   * Incremental cache check. Returns true if file hasn't changed.
   */
  isFileCached(fileId: string, contentHash: string, parserVersion: string, indexerVersion: string): boolean {
    const entry = this.fileCache.get(fileId)
    if (!entry) return false
    return (
      entry.contentHash === contentHash &&
      entry.parserVersion === parserVersion &&
      entry.indexerVersion === indexerVersion
    )
  }

  /**
   * Evicts a file and all symbols defined by it from the graph incrementally.
   */
  removeFile(fileId: string): string[] {
    const removedNodeIds = new Set<string>([fileId])

    // Find all nodes defined by this file
    for (const [id, node] of this.nodes) {
      if (node.fileId === fileId) {
        removedNodeIds.add(id)
      }
    }

    // Delete nodes
    for (const id of removedNodeIds) {
      this.nodes.delete(id)
      this.edgesFromMap.delete(id)
      this.edgesToMap.delete(id)
    }

    // Rebuild edges array and indices omitting edges touching removed nodes
    const survivingEdges = this.edges.filter((e) => !removedNodeIds.has(e.from) && !removedNodeIds.has(e.to))
    this.edges.length = 0
    this.edges.push(...survivingEdges)

    this.rebuildEdgeMaps()
    this.fileCache.delete(fileId)
    this.graphRevision = this.graphRevision.next()

    return Array.from(removedNodeIds)
  }

  /**
   * Records cache entry for an indexed file.
   */
  recordFileIndex(
    fileId: string,
    contentHash: string,
    parserVersion: string,
    indexerVersion: string,
    nodeIds: readonly string[]
  ): void {
    this.fileCache.set(fileId, {
      contentHash,
      parserVersion,
      indexerVersion,
      nodeIds,
    })
    this.graphRevision = this.graphRevision.next()
  }

  private rebuildEdgeMaps(): void {
    this.edgesFromMap.clear()
    this.edgesToMap.clear()
    for (const edge of this.edges) {
      const fromList = this.edgesFromMap.get(edge.from) ?? []
      fromList.push(edge)
      this.edgesFromMap.set(edge.from, fromList)

      const toList = this.edgesToMap.get(edge.to) ?? []
      toList.push(edge)
      this.edgesToMap.set(edge.to, toList)
    }
  }
}
